core/base-service/redirector.spec.js

```diff
@@ -0,0 +1,95 @@
+'use strict'
+
+const Camp = require('camp')
+const got = require('got')
+const portfinder = require('portfinder')
+const { expect } = require('chai')
+const redirector = require('./redirector')
+
+describe('Redirector', function() {
+  const route = {
+    base: 'very/old/service',
+    pattern: ':namedParamA',
+  }
+
+  it('returns true on isDeprecated', function() {
+    const ServiceClass = redirector({ route })
+    expect(ServiceClass.isDeprecated).to.be.true
+  })
+
+  it('sets specified route', function() {
+    const ServiceClass = redirector({ route })
+    expect(ServiceClass.route).to.deep.equal(route)
+  })
+
+  it('sets specified category', function() {
+    const category = 'analysis'
+    const ServiceClass = redirector({ route, category })
+    expect(ServiceClass.category).to.equal(category)
+  })
+
+  describe('ScoutCamp integration', function() {
+    let port, baseUrl
+    beforeEach(async function() {
+      port = await portfinder.getPortPromise()
+      baseUrl = `http://127.0.0.1:${port}`
+    })
+
+    let camp
+    beforeEach(async function() {
+      camp = Camp.start({ port, hostname: '::' })
+      await new Promise(resolve => camp.on('listening', () => resolve()))
+    })
+    afterEach(async function() {
+      if (camp) {
+        await new Promise(resolve => camp.close(resolve))
+        camp = undefined
+      }
+    })
+
+    const target = ({ namedParamA }) => `/new/service/${namedParamA}`
+
+    beforeEach(function() {
+      const ServiceClass = redirector({ route, target })
+      ServiceClass.register({ camp })
+    })
+
+    it('should redirect as configured', async function() {
+      const { statusCode, headers } = await got(
+        `${baseUrl}/very/old/service/hello-world.svg`,
+        {
+          followRedirect: false,
+        }
+      )
+
+      expect(statusCode).to.equal(301)
+      expect(headers.location).to.equal('/new/service/hello-world.svg')
+    })
+
+    it('should preserve the extension', async function() {
+      const { statusCode, headers } = await got(
+        `${baseUrl}/very/old/service/hello-world.png`,
+        {
+          followRedirect: false,
+        }
+      )
+
+      expect(statusCode).to.equal(301)
+      expect(headers.location).to.equal('/new/service/hello-world.png')
+    })
+
+    it('should forward the query params', async function() {
+      const { statusCode, headers } = await got(
+        `${baseUrl}/very/old/service/hello-world.svg?colorB=123&style=flat-square`,
+        {
+          followRedirect: false,
+        }
+      )
+
+      expect(statusCode).to.equal(301)
+      expect(headers.location).to.equal(
+        '/new/service/hello-world.svg?colorB=123&style=flat-square'
+      )
+    })
+  })
```

Comment:
Is it feasible (or even helpful for that matter) to test the cached/304 scenario?

Reply:
The underlying functions are tested. 🤷‍♂️Seems like a test would be nice to have, not essential (mostly in comparison to other parts of the roadmap). Also I'd really like to replace scoutcamp with something more modern and anything that touches a request would likely need to be rewritten. I guess that's not an argument against testing, more contributing to an instinct to put a light touch on this part of the code. This is all adapted from `base-static.js` by the way.

Reply:
works for me